python/cudf/cudf/core/reshape.py

```diff
@@ -681,3 +683,212 @@ def merge_sorted(
     )
     result._copy_categories(objs[0])
     return result
+
+
+def _pivot(df, index, columns):
+    """
+    Reorganize the values of the DataFrame according to the given
+    index and columns.
+
+    Parameters
+    ----------
+    df : DataFrame
+    index : cudf.core.index.Index
+        Index labels of the result
+    columns : cudf.core.index.Index
+        Column labels of the result
+    """
+    columns_labels, columns_idx = columns._encode()
+    index_labels, index_idx = index._encode()
+    column_labels = columns_labels.to_pandas().to_flat_index()
+
+    # the result of pivot always has a multicolumn
+    result = cudf.core.column_accessor.ColumnAccessor(
+        multiindex=True, level_names=(None,) + columns._data.names
+    )
+
+    def as_tuple(x):
+        return x if isinstance(x, tuple) else (x,)
+
+    for v in df:
+        names = [as_tuple(v) + as_tuple(name) for name in column_labels]
+        col = df._data[v]
+        result.update(
+            cudf.DataFrame._from_table(
+                col.scatter_to_table(
+                    index_idx,
+                    columns_idx,
+                    names,
+                    nrows=len(index_labels),
+                    ncols=len(names),
+                )
+            )._data
+        )
+
+    return cudf.DataFrame(
+        result, index=cudf.Index(index_labels, name=index.name)
+    )
+
+
+def pivot(df, index=None, columns=None, values=None):
+    """
+    Return reshaped DataFrame organized by the given index and column values.
+
+    Reshape data (produce a "pivot" table) based on column values. Uses
+    unique values from specified `index` / `columns` to form axes of the
+    resulting DataFrame.
+
+    Parameters
+    ----------
+    index : column name, optional
+        Column used to construct the index of the result.
+    columns : column name, optional
+        Column used to construct the columns of the result.
+    values : column name or list of column names, optional
+        Column(s) whose values are rearranged to produce the result.
+        If not specified, all remaining columns of the DataFrame
+        are used.
+
+    Returns
+    -------
+    DataFrame
+
+    Examples
+    --------
+    >>> a = cudf.DataFrame()
+    >>> a['a'] = [1, 1, 2, 2],
+    >>> a['b'] = ['a', 'b', 'a', 'b']
+    >>> a['c'] = [1, 2, 3, 4]
+    >>> a.pivot(index='a', columns='b')
+       c
+    b  a  b
+    a
+    1  1  2
+    2  3  4
+
+    Pivot with missing values in result:
+
+    >>> a = cudf.DataFrame()
+    >>> a['a'] = [1, 1, 2]
+    >>> a['b'] = [1, 2, 3]
+    >>> a['c'] = ['one', 'two', 'three']
+    >>> a.pivot(index='a', columns='b')
+              c
+        b     1     2      3
+        a
+        1   one   two   <NA>
+        2  <NA>  <NA>  three
+
+    """
+    if values is None:
+        values = df._columns_view(
+            col for col in df.columns if col not in (index, columns)
```

Comment:
Can we check for it in something like `df._column_names`? `df.columns` converts to a Pandas index object which adds a bit of overhead.

Reply:
Done